src/hypothesis/strategies.py

```diff
@@ -1286,19 +1296,79 @@ def int_to_decimal(val):
     else:
         # Otherwise, they're like fractions featuring a power of ten
         def fraction_to_decimal(val):
-            precision = ceil(math.log10(abs(val.numerator) or 1) +
-                             math.log10(val.denominator)) + 1
-            return Context(prec=precision or 1).divide(
-                Decimal(val.numerator), val.denominator)
+            precision = ceil(Decimal(abs(val.numerator) or 1).adjusted() +
+                             Decimal(val.denominator).adjusted()) + 1
+            return Context(prec=precision).divide(
+                val.numerator, val.denominator)
+
+        # Round bounding values to at most ten thousand digits of precision,
+        # so that casting to Fraction doesn't eat too much time and memory
+        exact_bounds = dict(min=min_value, max=max_value)
+        ctx = Context(prec=10000)
+        if min_value is not None:
+            min_value = ctx.add(min_value, 0)
+            if exact_bounds['min'] < min_value:
+                min_value = ctx.next_plus(min_value)  # pragma: no cover
+        if max_value is not None:
+            max_value = ctx.add(max_value, 0)
+            if exact_bounds['max'] > max_value:
+                max_value = ctx.next_minus(max_value)
+        if None not in (min_value, max_value) and min_value > max_value:
+            raise InvalidArgument(
+                'There are no decimal values between the given minimum and '
+                'maximum values with at most ten thousand digits of precison.'
+            )
+
+        # We split this into several cases, so that we can handle very large
+        # exponents without the pathological time and memory usage
+        if min_value is None and max_value is None:
+            # The trivial case - just use default values.
+            strat = fractions().map(fraction_to_decimal)
+        elif max_value is None:
+            # With a single bound, we can scale so the exponent is zero,
+            # generate, and rescale the example.
+            scale = Decimal(10) ** min_value.adjusted()
+            strat = fractions(min_value=min_value / scale)\
+                .map(fraction_to_decimal).map(lambda d: d * scale)
+        elif min_value is None:
+            scale = Decimal(10) ** max_value.adjusted()
+            strat = fractions(max_value=max_value / scale)\
+                .map(fraction_to_decimal).map(lambda d: d * scale)
+        elif abs(min_value.adjusted()) < 99 and abs(max_value.adjusted()) < 99:
+            # If we have two bounds which are "not too large", the naive
+            # approach simplifies shrinking and handles the common use-cases.
```

Comment:
What's the problem with shrinking in the non-naive approach? What happens when that fires?

Reply:
I think I wrote this before the `bind_points` work, when drawing a scale factor and then drawing everything else risked (relative) performance issues in shrinking.

Now, it's just trading off between one big draw and multiple linked smaller draws, and 99 digits seemed as good as any other limit.